### PR TITLE
do not follow symlink if it is preserved

### PR DESCRIPTION
**Pull request by atetubou**
This causes deadlock when there are cycles.

This is for https://crbug.com/1237607#c5

### CODE REVIEW

File: go/pkg/cas/upload.go [hunk 1]
```diff
@@ -810,24 +806,11 @@ func (u *uploader) visitSymlink(ctx context.Context, absPath string, pathExclude
 		return &digested{dirEntry: symlinkNode}, nil
 	case err != nil:
 		return nil, errors.WithStack(err)
+	case u.PreserveSymlinks:
```
Comment (Jwata) [inline go/pkg/cas/upload.go:809]: The case at L808 isn't necessary anymore?

**Review by Jwata — APPROVED** [review]

**Review by rubensf — COMMENTED** [review]
I'm not sure this is the correct thing to do here - if you don't follow the symlink, how's the symlink going to be used by the remote action? Also:

* The tools we run for automatic dependency resolution don't follow symlinks. This is arguably not the SDK's problem.
* This becomes somewhat inconsistent (depending on your point of view) with the non-preserve symlink option that follows symlinks.
* Particularly, this becomes inconsistent with the other upload implementation.

My points are somewhat arbitrary on what I _think_ the SDK should be responsible for, so feel free to counter argue. But I'd rather the solution here be either a) Add sanity check code that complains about loops; or b) Add an configuration option for "don't follow preserved symlinks",

**Comment by nodirg** [comment]
@atetubou did you consider making `visitPath` resilient to re-entrance? For example, if a path is being visited (cycle), then don't attempt to start processing it again (do not enter the cycle). Yes, `visitSymlink` wouldn't be able to return digest in this case, but it is not defined in this case anyway.

**Review by nodirg — APPROVED** [review]
LGTM to unblock, but the behavior could be improved.

**Comment by atetubou** [comment]
I think this package currently supports use case that nobody needs.
And due to that this package has some unnecessary code/test, I think deferring input specification to caller can simplify the codebase even though it has inconsistency with legacy package.

For that reason, I rather want to drop unused functionalities from this package until when there is actual necessity of such features.

So
* Do you think cas package will be used by other than chromium client near future?
* If so, following symlink is necessary for those use case (e.g. cannot specify target of symlink from caller) and this package needs to have option controlling the behavior?

I currently feel existing options are [YAGNI](http://go/tott/539).

**Review by rubensf — APPROVED** [review]
TBH, you're right I don't think we have any other users besides chromium for the cas package :) But I did have people ask me about the Preserve option a couple times before. Also, part of the caveat of having this open-source is that we may have unknown users.

We do hope to make the cas package the default long term, and maintaining consistency is useful for that. That said, this isn't going to happen anytime soon, so also approving for the sake of unblocking - just check nodir's comments about documenting and leaving a TODO.

File: go/pkg/cas/upload.go [hunk 2]
```diff
@@ -803,31 +799,21 @@ func (u *uploader) visitSymlink(ctx context.Context, absPath string, pathExclude
...
-		// resulting in a dangling symlink, which is not allowed.
-		return nil, errors.Wrapf(ErrFilteredSymlinkTarget, "path: %q, target: %q", absPath, target)
-	default:
-		ret := &digested{dirEntry: symlinkNode}
-		if digestedTarget != nil {
-			ret.digest = digestedTarget.digest
-		}
-		return ret, nil
+	if u.PreserveSymlinks {
+		return &digested{dirEntry: symlinkNode}, nil
```
[the remark this answers is not in the document]
Comment (atetubou) [reply go/pkg/cas/upload.go:815]: Added.

**Comment by atetubou** [comment]
Thank you for approval, I'll submit this.

I'm still wondering the possibility to move responsibility of symlink input specification to caller, or when it becomes problematic.
Though I may not take care implementation if I only need to file feature requests or bug reports and those are handled quickly for our use csae. But currently this package is practically somewhat owned by Chromium folks, so implementation comes to my mind.

**Comment by rubensf** [comment]
FWIW - thanks a lot for sending all these fixes :)

**Pull request merged**